### PR TITLE
[#9] Set up project app flavors

### LINKED ISSUE #9: [Android] Setup project app flavors
## Why

We must prepare 2 app build flavors: staging and production for the Android app. These app build flavors help us to config and distribute corresponding apps on each staging & production API environment.

## Acceptance Criteria

- Set up project app flavors: `staging` and `production`.
- Set up app names.
- Set up app icons.

### PR DESCRIPTION
**Pull request by lydiasama**
- Close #9 

## What happened 👀

- Added `Production` and `Staging` app flavors
- Added `release` and `debug` build type
- Generated keystone for both `release` and `debug`
- Configured app name for staging and production flavor

## Insight 📝

The `release.keystore` and `signing.properties` are provided in 1P [here](https://start.1password.com/open/i?a=4UPGXNITPJF2XPHOHZF57GRDVY&v=rlcs67pyv5umqcddmy2gutkfni&i=7nekb2ukakvqyc6mlswurj5ktq&h=nimble.1password.com).

## Proof Of Work 📹

SCR-25660804-ntzf

### CODE REVIEW

File: android/build.gradle.kts [hunk 1]
```diff
@@ -13,6 +17,22 @@ android {
...
+            storePassword = keystoreProperties.getProperty("KEYSTORE_PASSWORD")
+            // Remember to edit signing.properties to have the correct info for release build.
+            storeFile = file("config/release.keystore")
+            keyPassword = keystoreProperties.getProperty("KEY_PASSWORD")
+            keyAlias = keystoreProperties.getProperty("KEY_ALIAS")
+        }
+
+        getByName("debug") {
+            storeFile = file("config/debug.keystore")
+            storePassword = "jsglOe09kgjslDgpEg"
```
Comment (phongvhd93) [inline android/build.gradle.kts]: Is the value sensitive? If true, please store it in a secure place.

Comment (lydiasama) [reply android/build.gradle.kts:31]: @phongvhd93 No. It's just a debug build. In Android Template we do this too. 🙏🏻

File: android/build.gradle.kts [hunk 2]
```diff
@@ -1,8 +1,12 @@
+import org.jetbrains.kotlin.konan.properties.loadProperties
+
 plugins {
     id("com.android.application")
     kotlin("android")
 }
 
+val keystoreProperties = loadProperties("android/signing.properties")
```
Comment (suho) [inline android/build.gradle.kts]: @kaungkhantsoe You might need to update your code: https://github.com/nimblehq/kmm-templates/pull/25/files#diff-74804964ccc306f732ea790fdd0a41b45f4d6a0426c9305a7c91e900e1303116R4

File: android/build.gradle.kts (hunk 1 above)
Comment (suho) [reply android/build.gradle.kts]: @phongvhd93 For debug, it's ok to put these values here

**Review by suho — APPROVED** [review]

**Comment by luongvo** [comment]
This PR should be based on https://github.com/nimblehq/kmm-templates/pull/28 to verify the generated project changes.

File: android/build.gradle.kts (hunk 1 above)
Comment (markgravity) [reply android/build.gradle.kts]: At least we need to put some comments here. Ex: `Replace it with your password` 🙏

**Comment by phongvhd93** [comment]
@lydiasama please resolve the conflict

**Review by phongvhd93 — APPROVED** [review]

**Comment by lydiasama** [comment]
@phongvhd93 Done. 🙏🏻

File: sample/ios/ArkanaKeys/ArkanaKeys/Sources/ArkanaKeys.swift [hunk 3]
```diff
@@ -7,7 +7,7 @@ import ArkanaKeysInterfaces
 public enum ArkanaKeys {
     @inline(__always)
     fileprivate static let salt: [UInt8] = [
-        0x5d, 0xa0, 0x4d, 0x30, 0x47, 0xa6, 0xc3, 0x36, 0xda, 0x7c, 0xd5, 0xfd, 0xc6, 0xa, 0xfd, 0x98, 0xa5, 0x79, 0x88, 0x6c, 0x73, 0x63, 0x19, 0x95, 0x5d, 0x3a, 0xc8, 0x4, 0x9e, 0xd8, 0x8f, 0x26, 0x3f, 0x77, 0x39, 0xa2, 0xfa, 0xfb, 0x77, 0x3a, 0x4c, 0x2c, 0x53, 0x60, 0xba, 0x2f, 0xec, 0x43, 0x1f, 0x9f, 0xb, 0x4f, 0x9b, 0x88, 0xa6, 0xdd, 0xfc, 0x10, 0x3b, 0x3e, 0x1b, 0, 0x86, 0x72
+        0xc4, 0xf5, 0x2e, 0x24, 0x60, 0xf6, 0x94, 0xe6, 0x41, 0x28, 0x4d, 0x82, 0x95, 0x8d, 0xf, 0xdb, 0x35, 0x3c, 0xbf, 0x49, 0xd6, 0xbf, 0xa1, 0xe5, 0x32, 0xb5, 0x37, 0x9a, 0x4b, 0x48, 0xff, 0xf2, 0x71, 0x22, 0x9c, 0xa8, 0x44, 0x2c, 0x6d, 0xe6, 0xef, 0x9a, 0x46, 0xd1, 0x32, 0x31, 0xb3, 0xae, 0x3a, 0x42, 0x1e, 0x7e, 0x5d, 0xb, 0x6d, 0xa2, 0x5, 0x2c, 0xcb, 0x27, 0x9e, 0xe, 0x85, 0xde
```
Comment (luongvo) [inline sample/ios/ArkanaKeys/ArkanaKeys/Sources/ArkanaKeys.swift]: @phongvhd93 FYI: we should consider skipping Arkana running when generating the updated sample project here https://github.com/nimblehq/kmm-templates/blob/4fbe1f0c7e92c910db5f0564dc7c0671ad0aecf3/.github/workflows/test.yml#L35

File: android/config/debug.keystore
Comment (luongvo) [inline android/config/debug.keystore:1]: `android` is the app module folder of Android. We should follow the put the key out of the app folder https://github.com/nimblehq/android-templates/tree/develop/template-compose/config.

image


Actually, I put it here already https://github.com/nimblehq/kmm-templates/tree/develop/config

File: android/build.gradle.kts [hunk 4]
```diff
@@ -1,9 +1,11 @@
+import org.jetbrains.kotlin.konan.properties.loadProperties
+
 plugins {
     id(Plugins.ANDROID_APPLICATION)
     kotlin(Plugins.ANDROID)
 }
 
-val keystoreProperties = rootDir.loadGradleProperties("signing.properties")
+val keystoreProperties = loadProperties("android/signing.properties")
```
Comment (luongvo) [inline android/build.gradle.kts]: Good replacement, can we remove `loadGradleProperties` and `FileExt`?

File: android/.gitignore [hunk 5]
```diff
@@ -2,3 +2,4 @@
 google-services.json
 # Keystore
 config/release.keystore
+signing.properties
```
Comment (luongvo) [inline android/.gitignore]: The file should be at [the root](https://github.com/nimblehq/kmm-templates/blob/develop/signing.properties). Please skip it from root .gitignore 🙏

File: android/build.gradle.kts [hunk 6]
```diff
@@ -57,6 +59,14 @@ android {
             signingConfig = signingConfigs[BuildTypes.DEBUG]
         }
     }
+    flavorDimensions += "version"
```
Comment (luongvo) [inline android/build.gradle.kts]:
```suggestion
    flavorDimensions += Flavors.DIMENSION_VERSION
```

File: android/build.gradle.kts [hunk 7]
```diff
@@ -57,6 +59,14 @@ android {
...
+    productFlavors {
+        create("staging") {
```
Comment (luongvo) [inline android/build.gradle.kts]:
```suggestion
        create(Flavors.STAGING) {
```

File: android/build.gradle.kts [hunk 8]
```diff
@@ -57,6 +59,14 @@ android {
...
+            applicationIdSuffix = ".staging"
+        }
+
+        create("production") {}
```
Comment (luongvo) [inline android/build.gradle.kts]:
```suggestion
        create(Flavors.PRODUCTION) {}
```

File: android/build.gradle.kts (hunk 7 above)
Comment (luongvo) [inline android/build.gradle.kts]: Could you please define the app name for [Staging](https://github.com/nimblehq/android-templates/blob/develop/template-compose/app/src/staging/res/values/strings.xml) and [Production](https://github.com/nimblehq/android-templates/blob/develop/template-compose/app/src/main/res/values/strings.xml)? We can skip defining app icons.

image

**Review by luongvo — DISMISSED** [review]

File: android/.gitignore (hunk 5 above)
Comment (lydiasama) [reply android/.gitignore:5]: @luongvo Are you saying that we should put the `signing.properties` on the root path instead inside of `android` package?

**Review by luongvo — APPROVED** [review]

**Pull request merged**